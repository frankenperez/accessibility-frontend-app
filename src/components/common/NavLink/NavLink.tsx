/* eslint-disable react/display-name */
// React
import React from "react";

// Router
import { NavLink as NavLinkReactRouter, NavLinkProps as NavLinkReactRouterProps } from "react-router-dom";

export interface NavLinkProps extends NavLinkReactRouterProps {
  activeClassName: string;
}

// Component wrapper to forward props to NavLink from React Router
const NavLink = React.forwardRef<HTMLAnchorElement, NavLinkProps>((props: NavLinkProps, ref) => {
  return (
    <NavLinkReactRouter
      ref={ref}
      to={props.to}
      className={({ isActive }) => `${props.className} ${isActive ? props.activeClassName : ""}`}
    >
      {props.children}
    </NavLinkReactRouter>
  );
});

export default NavLink;
