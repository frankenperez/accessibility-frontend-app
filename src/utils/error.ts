import axios from "axios";

export const errorParser = (error: unknown) => {
  let message: string;
  if (axios.isAxiosError(error)) {
    const errorData = error.response?.data as { reason: string } | string;
    if (typeof errorData === "object" && errorData?.reason) {
      message = errorData?.reason;
    } else {
      message = JSON.stringify(errorData);
    }
  } else if (error instanceof Error) {
    message = error.message;
  } else {
    message = JSON.stringify(error);
  }
  return message;
};
