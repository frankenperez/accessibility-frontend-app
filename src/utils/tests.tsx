// React
import React, { PropsWithChildren } from "react";

// Testing Library
import { render, RenderOptions } from "@testing-library/react";

// Router
import { BrowserRouter } from "react-router-dom";

// Internationalization
import en_BG from "../languages/en-GB.json";
import i18next from "i18next";
import { I18nextProvider } from "react-i18next";

// MUI Theme
import { createTheme, StyledEngineProvider, ThemeProvider } from "@mui/material/styles";
import theme from "styles/theme";
const appTheme = createTheme(theme);

// Internationalization config
i18next.init({
  interpolation: { escapeValue: false }, // React already does escaping
  lng: "en", // Language to use
  resources: {
    en: {
      app: en_BG
    }
  }
});

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  route?: string;
}

const renderWithProviders = (
  ui: React.ReactElement,
  {
    route = "/",
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => {
  const Wrapper = ({ children }: PropsWithChildren): JSX.Element => {
    return (
      <I18nextProvider i18n={i18next}>
        <StyledEngineProvider injectFirst>
          <ThemeProvider theme={appTheme}>
            <BrowserRouter>{children}</BrowserRouter>
          </ThemeProvider>
        </StyledEngineProvider>
      </I18nextProvider>
    );
  };
  window.history.pushState({}, "Test page", route);

  // Return an object with the store and all of RTL's query functions
  return { ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
};

export { renderWithProviders };
