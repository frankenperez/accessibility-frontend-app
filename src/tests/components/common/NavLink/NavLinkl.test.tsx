// React
import React from "react";

// Component
import NavLink, { NavLinkProps } from "components/common/NavLink/NavLink";

//Test Library
import { screen, fireEvent } from "@testing-library/react";

//Test Utils
import { renderWithProviders } from "utils/tests";

// Defaults
const defaultProps: NavLinkProps = {
  to: "/",
  activeClassName: "active"
};

describe("NavLink", () => {
  it("renders correctly", () => {
    const { container } = renderWithProviders(<NavLink {...defaultProps}>Link</NavLink>);

    expect(container).toMatchSnapshot();
  });

  it("changes className when active", () => {
    renderWithProviders(<NavLink {...defaultProps}>Link</NavLink>);

    const link = screen.getByText("Link");

    expect(link.className).toContain("active");
  });

  it("updates URL when clicked", async () => {
    const combinedProps = { ...defaultProps, to: "/new-page" };

    renderWithProviders(<NavLink {...combinedProps}>Link</NavLink>);
    const link = screen.getByText("Link");
    fireEvent.click(link);

    expect(window.location.pathname).toBe("/new-page");
  });
});
