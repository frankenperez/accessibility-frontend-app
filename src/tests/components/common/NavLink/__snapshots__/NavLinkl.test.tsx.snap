// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavLink renders correctly 1`] = `
<div>
  <a
    aria-current="page"
    class="undefined active"
    href="/"
  >
    Link
  </a>
</div>
`;
